pages/views/elements.py

```diff
@@ -218,3 +226,44 @@ def get_context_data(self, **kwargs):
                         kwargs=url_kwargs),
                 })
         return context
+
+
+class CertificateDownloadView(LoginRequiredMixin, DetailView):
+    model = Sequence
+    slug_url_kwarg = 'sequence_slug'
+    template_name = 'pages/certificate.html'
+    response_class = PdfTemplateResponse
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        sequence = self.object
+        sequence_progress = get_object_or_404(
+            SequenceProgress,
+            sequence=sequence,
+            user=self.request.user)
+        enumerated_elements = EnumeratedElements.objects.filter(
+            sequence=sequence)
+        user_enumerated_progress = EnumeratedProgress.objects.filter(
+            progress=sequence_progress)
+
+        has_completed_sequence = True
+        for element in enumerated_elements:
+            user_element_progress = user_enumerated_progress.filter(rank=element.rank).first()
+            if not user_element_progress or user_element_progress.viewing_duration < element.min_viewing_duration:
+                has_completed_sequence = False
+                break
+
+        if has_completed_sequence:
+            last_element = enumerated_elements.last().page_element
+            context.update({
+                'user': self.request.user,
+                'sequence': sequence,
+                'current_date': date.today(),
```

Comment:
Interesting. We rely on the current Web server date.
1. It could be important to pass the timezone as a query parameter just to make sure a one off day does not trigger e-mail and calls to tech support.
2. The date should most likely be fixed so that two successive downloads do not return different date. This might be important for CEUs and other certificate that expire, triggering taking the course again.